// projectz/out/lib/badge-util.js
// Generated by CoffeeScript 1.6.3
(function() {
  var badgeUtil,
    __hasProp = {}.hasOwnProperty;

  module.exports = badgeUtil = {
    miscBadges: {
      travis: function(opts) {
        if (opts == null) {
          opts = {};
        }
        if (!opts.badges.travis || !opts.username || !opts.name) {
          return '';
        }
        return "[![Build Status](http://img.shields.io/travis-ci/" + opts.username + "/" + opts.name + ".png?branch=master)](http://travis-ci.org/" + opts.username + "/" + opts.name + " \"Check this project's build status on TravisCI\")";
      },
      npm: function(opts) {
        if (opts == null) {
          opts = {};
        }
        if (!opts.badges.npm || !opts.name) {
          return '';
        }
        return "[![NPM version](http://badge.fury.io/js/" + opts.name + ".png)](https://npmjs.org/package/" + opts.name + " \"View this project on NPM\")";
      }
    },
    donationBadges: {
      gittip: function(opts) {
        var url;
        if (opts == null) {
          opts = {};
        }
        if (!opts.badges.gittip) {
          return '';
        }
        url = "https://www.gittip.com/" + opts.badges.gittip + "/";
        return "[![Gittip donate button](http://img.shields.io/gittip/" + opts.badges.gittip + ".png)](" + url + " \"Donate weekly to this project using Gittip\")";
      },
      flattr: function(opts) {
        var url;
        if (opts == null) {
          opts = {};
        }
        if (!opts.badges.flattr) {
          return '';
        }
        url = "http://flattr.com/thing/" + opts.badges.flattr;
        return "[![Flattr donate button](http://img.shields.io/flattr/donate.png?color=yellow)](" + url + " \"Donate monthly to this project using Flattr\")";
      },
      paypal: function(opts) {
        var url;
        if (opts == null) {
          opts = {};
        }
        if (!opts.badges.paypal) {
          return '';
        }
        url = "https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=" + opts.badges.paypal;
        return "[![PayPayl donate button](http://img.shields.io/paypal/donate.png?color=yellow)](" + url + " \"Donate once-off to this project using Paypal\")";
      }
    },
    getBadgesSection: function(opts) {
      if (opts == null) {
        opts = {};
      }
      return badgeUtil.getTypedBadges('misc', opts) + '\n' + badgeUtil.getTypedBadges('donation', opts);
    },
    getTypedBadges: function(type, opts) {
      var fn, name, result, results, _ref;
      if (opts == null) {
        opts = {};
      }
      results = [];
      _ref = badgeUtil[type + 'Badges'];
      for (name in _ref) {
        if (!__hasProp.call(_ref, name)) continue;
        fn = _ref[name];
        result = fn.call(this, opts);
        if (result) {
          results.push(result);
        }
      }
      return results.join('\n');
    }
  };

}).call(this);
