// projectz/out/lib/license-util.js
// Generated by CoffeeScript 1.6.3
(function() {
  var licenseUtil, projectzUtil;

  projectzUtil = require('./projectz-util');

  module.exports = licenseUtil = {
    getAuthorText: function(author) {
      if (author == null) {
        author = {};
      }
      return "Copyright &copy; " + (projectzUtil.getPersonText(author));
    },
    getLicenseText: function(license) {
      var fn;
      fn = licenseUtil.sections[license.type.toLowerCase()];
      if (fn) {
        return fn.call(this, license);
      } else {
        return licenseUtil.sections.unknown.call(this, license);
      }
    },
    getLicenseFileText: function(license) {
      var fn;
      fn = licenseUtil.files[license.type.toLowerCase()];
      if (fn) {
        return fn.call(this, license);
      } else {
        return licenseUtil.files.unknown.call(this, license);
      }
    },
    getAuthors: function(opts) {
      var author, result;
      result = '';
      if (opts.authors.length === 0) {

      } else if (opts.authors.length === 1) {
        author = opts.authors[0];
        result += licenseUtil.getAuthorText(author);
      } else {
        result += ((function() {
          var _i, _len, _ref, _results;
          _ref = opts.authors;
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            author = _ref[_i];
            _results.push(licenseUtil.getAuthorText(author));
          }
          return _results;
        })()).join('\n<br/>');
      }
      return result;
    },
    getLicenses: function(opts) {
      var license, result;
      result = '';
      if (opts.licenses.length === 0) {

      } else if (opts.licenses.length === 1) {
        license = opts.licenses[0];
        result += "Licensed under " + (licenseUtil.getLicenseText(license));
      } else {
        result += "Licensed under:\n\n";
        result += '- ' + ((function() {
          var _i, _len, _ref, _results;
          _ref = opts.licenses;
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            license = _ref[_i];
            _results.push(licenseUtil.getLicenseText(license));
          }
          return _results;
        })()).join('\n- ');
      }
      return result;
    },
    getLicensesFiles: function(opts) {
      var license;
      return ((function() {
        var _i, _len, _ref, _results;
        _ref = opts.licenses;
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          license = _ref[_i];
          _results.push(licenseUtil.getLicenseFileText(license));
        }
        return _results;
      })()).join('\n\n');
    },
    files: {
      mit: function(opts) {
        if (opts == null) {
          opts = {};
        }
        return "## The MIT License\nPermission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.";
      },
      unknown: function(license) {
        return "## " + license.type + "\n" + (licenseUtil.sections.unknown.call(this, license));
      }
    },
    sections: {
      mit: function(license) {
        var url;
        url = license.url || 'http://creativecommons.org/licenses/MIT/';
        return "the incredibly [permissive](http://en.wikipedia.org/wiki/Permissive_free_software_licence) [MIT license](" + url + ")";
      },
      unknown: function(license) {
        if (license.url) {
          return "[" + license.type + "](" + license.url + ")";
        } else {
          return license.type;
        }
      }
    },
    getLicenseSection: function(opts) {
      var result;
      if (opts == null) {
        opts = {};
      }
      if (!opts.licenses) {
        return '';
      }
      result = "## License\n\n" + (licenseUtil.getLicenses(opts)) + "\n\n" + (licenseUtil.getAuthors(opts));
      return result;
    },
    getLicenseFile: function(opts) {
      var result;
      if (opts == null) {
        opts = {};
      }
      if (!opts.licenses) {
        return '';
      }
      result = "# License\n\n" + (licenseUtil.getAuthors(opts)) + "\n\n" + (licenseUtil.getLicensesFiles(opts));
      return result;
    }
  };

}).call(this);
